File: Algorithm/RadixSort.m
% 基数排序，输入数组Array为正整数，且位数均不大于d，非原址
function SortArray=RadixSort(Array,d)
SortArray=Array;
len=length(Array);
Arrorder=zeros(1,len);
for i=1:d
    remainder=rem(Array,10);
    order=sub_countsort(remainder,9);
    for j=1:len
        Arrorder(j)=find(order==j);
    end
    SortArray=SortArray(Arrorder);
    Array=floor(SortArray/10^i);
end
end

function order=sub_countsort(Array,k)
len=length(Array);
CountArray=zeros(1,k+1);
for i=1:len
    CountArray(Array(i)+1)=CountArray(Array(i)+1)+1;
end
for i=2:k+1
    CountArray(i)=CountArray(i)+CountArray(i-1);
end
for i=len:-1:1
    order(i)=CountArray(Array(i)+1);
    CountArray(Array(i)+1)=CountArray(Array(i)+1)-1;
end
end

File: Algorithm/Sort & Order Statistic/C/RadixSort.c
//Ìí¼Óstdlib.h¼°math.h
void sub_countsort(int* A, int* order, int k, int arraysize)
{
	int* CountArray = (int*)malloc(sizeof(int) * (k + 1));
	for (int i = 0; i <= k; i++)
		*(CountArray + i) = 0;

	for (int i = 0; i < arraysize; i++)
		*(CountArray + *(A + i)) = *(CountArray + *(A + i)) + 1;
	for (int i = 1; i <= k; i++)
		*(CountArray + i) = *(CountArray + i) + *(CountArray + i - 1);
	for (int i = arraysize - 1; i >= 0; i--)
	{
		*(order + i) = *(CountArray + *(A + i)) - 1;
		*(CountArray + *(A + i)) = *(CountArray + *(A + i)) - 1;
	}
	free(CountArray);
}

void RadixSort(int* A, int* B, int d, int arraysize)
{
	int* tempArray = (int*)malloc(sizeof(int) * arraysize);
	int* order = (int*)malloc(sizeof(int) * arraysize);
	int* remainder = (int*)malloc(sizeof(int) * arraysize);
	int i, j, power;

	for (i = 0; i < arraysize; i++)
		*(tempArray + i) = *(A + i);

	for (i = 1; i <= d; i++)
	{
		for (j = 0; j < arraysize; j++)
			*(remainder + j) = *(A + j) % 10;
		sub_countsort(remainder, order, 9, arraysize);
		
		for (j = 0; j < arraysize; j++)
			*(B + *(order + j)) = *(tempArray + j);
		for (j = 0; j < arraysize; j++)
			*(tempArray + j) = *(B + j);

		power = (int)pow(10.0, i);
		for (j = 0; j < arraysize; j++)
			*(A + j) = *(tempArray + j) / power;
	}
	free(remainder);
	free(order);
	free(tempArray);
}